daemon/emer-daemon.c

```diff
@@ -1123,6 +1142,379 @@ set_max_bytes_buffered (EmerDaemon *self,
+  //TODO: buffer_all_aggregate_events_to_queue (self);
+  //TODO: clear_aggregate_events_tally (self);
+  schedule_next_month_tick (self);
+
+  return G_SOURCE_REMOVE;
+}
+
+static void
+schedule_next_midnight_tick (EmerDaemon *self)
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  g_autoptr (GDateTime) next_midnight = NULL;
+  g_autoptr (GDateTime) next_day = NULL;
+  g_autoptr (GDateTime) now = NULL;
+  gint64 time_to_next_midnight_us;
+
+  now = g_date_time_new_now_local ();
+  next_day = g_date_time_add_days (now, 1);
+  next_midnight = g_date_time_new_local (g_date_time_get_year (next_day),
+                                         g_date_time_get_month (next_day),
+                                         g_date_time_get_day_of_month (next_day),
+                                         0, 0, 0.0);
+
+  time_to_next_midnight_us = g_date_time_difference (next_midnight, now);
+
+  priv->dispatch_aggregate_timers_daily_source_id =
+    eins_boottimeout_add_useconds (time_to_next_midnight_us,
+                                   clock_ticked_midnight_cb,
+                                   self);
+
+  priv->current_aggregate_tally_date = g_date_time_ref (now);
+}
+
+static void
+schedule_next_month_tick (EmerDaemon *self)
+{
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+  g_autoptr (GDateTime) next_month_start = NULL;
+  g_autoptr (GDateTime) next_month = NULL;
+  g_autoptr (GDateTime) now = NULL;
+  gint64 time_to_next_month_us;
+
+  now = g_date_time_new_now_local ();
+  next_month = g_date_time_add_months (now, 1);
+  next_month_start =
+    g_date_time_new_local (g_date_time_get_year (next_month),
+                           g_date_time_get_month (next_month),
+                           1, 0, 0, 0.0);
+
+  time_to_next_month_us =
+    g_date_time_difference (next_month_start, now);
+
+  priv->dispatch_aggregate_timers_monthly_source_id =
+    eins_boottimeout_add_useconds (time_to_next_month_us,
+                                   clock_turned_month_cb,
+                                   self);
+}
+
+static void
+stop_aggregate_timer (AggregateTimerData *timer_data)
+{
+  timer_data->end_monotonic_us = g_get_monotonic_time ();
+
+  save_aggregate_timer_to_tally (timer_data);
+
+  g_clear_pointer (&timer_data->aggregate_key, g_variant_unref);
+  g_clear_pointer (&timer_data->cache_key, g_variant_unref);
+  g_clear_pointer (&timer_data->event_id, g_variant_unref);
+  g_clear_pointer (&timer_data->payload, g_variant_unref);
+  g_free (timer_data);
+}
+
+static gboolean
+on_timer_stopped_cb (EmerAggregateTimer    *timer,
+                     GDBusMethodInvocation *invocation,
+                     AggregateTimerData    *timer_data)
+{
+  EmerDaemon *self = timer_data->daemon;
+  EmerDaemonPrivate *priv = emer_daemon_get_instance_private (self);
+
+  // Removing from the hash table will call stop_aggregate_timer()
+  g_hash_table_remove (priv->aggregate_timers, timer_data->cache_key);
```

Comment:
I believe that after this point 'timer_data' will have been freed by the hash table? Perhaps worth a note to that effect.